File: ios/FreakCore/Sources/FreakCore/Session.swift
// Session.swift — FreakSession: seq counter, one transaction at a time, the
// reply-lag chokepoint (port of session.py).
//
// - The addressed-request seq counter is owned here; it increments 1..127
//   and wraps, never emitting 0, matching mfcap.midi.Reader. The write burst
//   has a separate seq stream, verbatim from the captures: goFrame() carries
//   seq 0 and the chunks continue from it, (i + 1) % 128 — wrapping THROUGH
//   0 (Wire.chunkFrames owns that stream).
// - FreakSession is an actor. Actor isolation gives data-race freedom, but
//   actor REENTRANCY would let a second caller interleave at any `await`
//   inside a write burst — exactly the catastrophe the Python lock prevents
//   (chunk streams are unaddressed and unmatchable by design). So every
//   public method runs its body inside a non-reentrant FIFO gate: "one
//   transaction at a time" is enforced by construction.
// - transactAddressed() is the ONLY function in the package that inspects a
//   0x52 reply — readName, writeName's read-back, and writePreset's frames
//   1 and 7 all pass through it, so the reply-lag bug cannot be reintroduced
//   without deleting this function.

import Foundation

public struct SessionConfig: Sendable {
    public var nameTimeout: TimeInterval = 1.0
    public var dumpTimeout: TimeInterval = 1.5
    public var ackTimeout: TimeInterval = 1.0
    public var nameRetries: Int = 3
    public var pollInterval: TimeInterval = 0.002   // sleep between empty receive polls

    public init() {}
}

public actor FreakSession {
    private let transport: any FreakTransport
    private let config: SessionConfig
    private let clock: any FreakClock

    /// Addressed-request seq counter: `seq = seq % 127 + 1` -> 1..127,
    /// wrapping, NEVER 0. (0 appears on the wire only in the go frame; the
    /// chunk stream is Wire.chunkFrames'.)
    private var seq = 0

    // Non-reentrant FIFO gate.
    private var busy = false
    private var waiters: [CheckedContinuation<Void, Never>] = []

    public init(transport: any FreakTransport,
                config: SessionConfig = SessionConfig(),
                clock: any FreakClock = SystemClock()) {
        self.transport = transport
        self.config = config
        self.clock = clock
    }

    // ------------------------------------------------------------- public

    public func readName(slot: Int) async throws -> NameInfo {
        try await withTransaction {
            try await self.transactAddressed(
                request: Wire.readNameRequest(seq: self.nextSeq(), slot: slot),
                slot: slot)
        }
    }

    public func readBlob(slot: Int) async throws -> Data {
        try await withTransaction {
            try await self.readBlobInTransaction(slot: slot)
        }
    }

    /// The gate-verified 7-frame write sequence, verbatim from
    /// docs/write-protocol.md. No checksum is computed at any point; none
    /// exists. Returns frame 7's read-back NameInfo — comparison is the
    /// caller's job (the frame itself is protocol fidelity and always sent).
    /// Polls Task.isCancelled before each chunk; throws
    /// .operationCancelled(done: chunkIndex, total: 146) — tearing the slot.
    public func writePreset(slot: Int, preset: Preset) async throws -> NameInfo {
        // Built first: validates size / 7-bit cleanliness up front, outside
        // the gate.
        let frames = try Wire.chunkFrames(blob: preset.blob)
        return try await withTransaction {
            try await self.writePresetInTransaction(slot: slot, preset: preset,
                                                    frames: frames)
        }
    }

    /// Rename-in-place, exactly what MCC sends: the long 0x52 (awaiting the
    /// device's 0x18 ack; absence -> .deviceTimeout(stage: .nameWriteAck))
    /// plus a refresh read. No blob traffic.
    public func writeName(slot: Int, name: String, meta: Data) async throws -> NameInfo {
        try await withTransaction {
            try await self.transport.send(Wire.nameWriteFrame(
                seq: self.nextSeq(), slot: slot, name: name, meta: meta))
            guard try await self.awaitAck() else {
                throw FreakError.deviceTimeout(stage: .nameWriteAck, slot: slot)
            }
            return try await self.transactAddressed(
                request: Wire.readNameRequest(seq: self.nextSeq(), slot: slot),
                slot: slot)
        }
    }

    public func close() async {
        await transport.close()
    }

    // -------------------------------------------------- the transaction gate

    /// Non-reentrant FIFO gate: acquire() suspends while busy (FIFO resume
    /// order); release() resumes the head waiter. Cancellation while queued:
    /// the waiter still acquires, then the op's first Task.isCancelled poll
    /// exits it promptly.
    private func withTransaction<T: Sendable>(
        _ body: () async throws -> T) async rethrows -> T {
        await acquire()
        defer { release() }
        return try await body()
    }

    private func acquire() async {
        if !busy {
            busy = true
            return
        }
        await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
            waiters.append(cont)
        }
        // Resumed by release() with busy left true for us.
    }

    private func release() {
        if waiters.isEmpty {
            busy = false
        } else {
            waiters.removeFirst().resume()
        }
    }

    // ------------------------------------------------------------ machinery

    private func nextSeq() -> UInt8 {
        seq = seq % 127 + 1     // 1..127, never 0
        return UInt8(seq)
    }

    private func drain() async throws {
        while try await transport.receive(timeout: 0) != nil {}
    }

    /// Next parseable MicroFreak frame within `within` seconds, else nil.
    /// An empty poll sleeps `pollInterval` on the injected clock (the sim
    /// returns nil immediately — this is what keeps the poll from spinning
    /// hot). A sleep interrupted by task cancellation returns early; callers
    /// re-check deadlines and the cancellation poll points exit promptly.
    private func nextFrame(within: TimeInterval) async throws -> Wire.Frame? {
        let deadline = clock.now + within
        while true {
            let remaining = deadline - clock.now
            if remaining <= 0 {
                return nil
            }
            guard let raw = try await transport.receive(timeout: remaining) else {
                try? await clock.sleep(for: config.pollInterval)
                continue
            }
            if let f = Wire.parse(raw) {
                return f
            }
        }
    }

    /// THE reply-lag chokepoint — the only code in the package that inspects
    /// a 0x52 reply. Send an addressed request; accept only a long-0x52
    /// reply whose embedded bank/pos names the requested slot.
    ///
    /// A mismatched reply is discarded as stale and the request resent,
    /// nameRetries attempts total; then .replyMismatch. Total silence throws
    /// .deviceTimeout(stage: .nameRead, slot:).
    ///
    /// Known limitation (verbatim from the Python): when two consecutive
    /// reads target the SAME slot (writePreset frames 1 and 7; a rename's
    /// read-back), a lagged reply to the first is indistinguishable from the
    /// answer to the second — the embedded address matches either way.
    /// Blob-hash verification covers writes; renames retain the window. No
    /// protocol-level fix exists — replies carry no request id. Do not
    /// "improve" this.
    private func transactAddressed(request: Data, slot: Int) async throws -> NameInfo {
        try await drain()
        var sawReply = false
        var repliedSlot: Int? = nil
        for _ in 0..<config.nameRetries {
            try await transport.send(request)
            let deadline = clock.now + config.nameTimeout
            scan: while true {
                let remaining = deadline - clock.now
                if remaining <= 0 {
                    break scan                  // silent attempt; resend
                }
                guard let f = try await nextFrame(within: remaining) else {
                    break scan                  // silent attempt; resend
                }
                if f.cmd != Wire.cmdName || f.data.count != Wire.namePayloadLength {
                    continue scan               // unrelated traffic; ignore
                }
                let info: NameInfo
                do {
                    info = try Wire.decodeNameReply(f)
                } catch let e as FreakError where e.group == .protocolError {
                    continue scan               // malformed (e.g. out-of-range
                                                // embedded address); ignore
                }
                sawReply = true
                if info.slot == slot {
                    return info
                }
                repliedSlot = info.slot         // stale (lagged) reply
                break scan                      // discard and resend at once
            }
        }
        if sawReply {
            throw FreakError.replyMismatch(requestedSlot: slot,
                                           repliedSlot: repliedSlot,
                                           attempts: config.nameRetries)
        }
        throw FreakError.deviceTimeout(stage: .nameRead, slot: slot)
    }

    /// Dump state machine: open, then pull chunks in strict lockstep — one
    /// outstanding request — until the device says "last".
    private func readBlobInTransaction(slot: Int) async throws -> Data {
        try await drain()
        try await transport.send(Wire.openDumpRequest(seq: nextSeq(), slot: slot))
        var chunks: [Wire.Frame] = []
        while true {
            // Sweep anything already delivered before pulling again; the
            // count bound keeps a runaway device (streaming 0x16 forever with
            // no terminator) from spinning this loop unboundedly.
            var gotLast = false
            while chunks.count < Wire.chunkCount {
                guard let raw = try await transport.receive(timeout: 0) else { break }
                guard let f = Wire.parse(raw), f.isChunk else { continue }
                chunks.append(f)
                if f.isLastChunk {
                    gotLast = true
                    break
                }
            }
            if gotLast {
                return try Wire.assembleBlob(chunks)
            }
            if chunks.count >= Wire.chunkCount {
                throw FreakError.protocolViolation(
                    detail: "slot \(slot): \(chunks.count) chunks without a 0x17 terminator")
            }
            try await transport.send(Wire.pullNextRequest(seq: nextSeq()))
            let f = try await awaitChunk(slot: slot)
            chunks.append(f)
            if f.isLastChunk {
                return try Wire.assembleBlob(chunks)
            }
        }
    }

    /// The 7-frame sequence with ack accounting (inside the gate).
    private func writePresetInTransaction(slot: Int, preset: Preset,
                                          frames: [Data]) async throws -> NameInfo {
        // 1. name read (fidelity to MCC; result discarded). Its own errors
        //    propagate unwrapped.
        _ = try await transactAddressed(
            request: Wire.readNameRequest(seq: nextSeq(), slot: slot), slot: slot)
        var chunksSent = 0
        var stage = WriteStage.nameWrite
        do {
            // 2. name + meta — the device acks it with 0x18 (captured: every
            //    outbound 0x52 and 0x15 is acked, not just chunks)
            try await transport.send(Wire.nameWriteFrame(
                seq: nextSeq(), slot: slot, name: preset.name, meta: preset.meta))
            try await requireAck(stage: .nameWrite, slot: slot, chunksSent: chunksSent)
            // 3. open blob write — acked with 0x18
            stage = .open
            try await transport.send(Wire.openWriteFrame(seq: nextSeq(), slot: slot))
            try await requireAck(stage: .open, slot: slot, chunksSent: chunksSent)
            // 4. go — acked with 0x18
            stage = .go
            try await transport.send(Wire.goFrame())
            try await requireAck(stage: .go, slot: slot, chunksSent: chunksSent)
            // 5-6. 146 chunks, each acked by the device with 0x18. The three
            // control-frame acks above are consumed before the chunk loop, so
            // ack N here really is the ack for chunk N — flow control stays
            // in lockstep with the device (the device acks EVERY write frame;
            // a writer pairing acks with chunks only would run three ahead).
            stage = .chunk
            for (i, ch) in frames.enumerated() {
                if Task.isCancelled {
                    // Cancelling mid-write tears the slot; the fields say how
                    // far it got — recovery is "write again".
                    throw FreakError.operationCancelled(done: i, total: frames.count)
                }
                try await transport.send(ch)
                chunksSent += 1
                if !(try await awaitAck()) {
                    throw FreakError.chunkNotAcked(slot: slot, chunkIndex: i)
                }
            }
        } catch let e as FreakError {
            switch e {
            case .chunkNotAcked, .operationCancelled, .writeAborted:
                throw e
            default:
                if e.group == .transport || e.group == .protocolError {
                    throw FreakError.writeAborted(stage: stage, slot: slot,
                                                  chunksSent: chunksSent)
                }
                throw e
            }
        }
        // 7. name read back
        do {
            return try await transactAddressed(
                request: Wire.readNameRequest(seq: nextSeq(), slot: slot), slot: slot)
        } catch let e as FreakError {
            switch e.group {
            case .transaction, .transport:
                throw FreakError.writeAborted(stage: .finalRead, slot: slot,
                                              chunksSent: chunksSent)
            default:
                throw e
            }
        }
    }

    /// Await the device's 0x18 ack for a write control frame (long 0x52,
    /// short 0x52 open, go); absence aborts the write at that stage.
    private func requireAck(stage: WriteStage, slot: Int, chunksSent: Int) async throws {
        if !(try await awaitAck()) {
            throw FreakError.writeAborted(stage: stage, slot: slot,
                                          chunksSent: chunksSent)
        }
    }

    private func awaitChunk(slot: Int) async throws -> Wire.Frame {
        let deadline = clock.now + config.dumpTimeout
        while true {
            let remaining = deadline - clock.now
            if remaining <= 0 {
                throw FreakError.deviceTimeout(stage: .dump, slot: slot)
            }
            guard let f = try await nextFrame(within: remaining) else {
                throw FreakError.deviceTimeout(stage: .dump, slot: slot)
            }
            if f.isChunk {
                return f
            }
            // anything else during a dump is stale; discard
        }
    }

    /// Device 0x18 within ackTimeout; false on timeout.
    private func awaitAck() async throws -> Bool {
        let deadline = clock.now + config.ackTimeout
        while true {
            let remaining = deadline - clock.now
            if remaining <= 0 {
                return false
            }
            guard let f = try await nextFrame(within: remaining) else {
                return false
            }
            if f.isAck {
                return true
            }
            // anything else during a write is stale; discard
        }
    }
}

File: microfreak/session.py
"""Session: seq counter, one transaction at a time, the reply-lag chokepoint.

Semantics (docs/core-api.md is the normative statement):

- The addressed-request seq counter is owned here; it increments 1..127 and
  wraps, never emitting 0, matching mfcap.midi.Reader. The write burst has a
  separate seq stream, verbatim from the captures: go_frame() carries seq 0
  and the chunks continue from it, (i + 1) % 128 — wrapping THROUGH 0
  (chunk_frames owns that stream).
- One internal threading.Lock wraps every public method, so concurrent
  callers queue instead of interleaving chunk streams (chunks are unaddressed
  and unmatchable by design; interleaving is forbidden, and the lock makes
  accidental interleaving impossible from one Session). One Session per
  transport.
- _transact_addressed() is the ONLY function in the package that inspects a
  0x52 reply — read_name, write_name's read-back, and write_preset's frames
  1 and 7 all pass through it, so the reply-lag bug cannot be reintroduced
  without deleting this function.
"""
from __future__ import annotations

import threading
import time
from typing import Callable, List, Optional

from . import protocol
from .errors import (ChunkNotAckedError, DeviceTimeoutError,
                     OperationCancelledError, ProtocolError,
                     ReplyMismatchError, TransportError, WriteAbortedError)
from .model import CancelToken, Preset
from .protocol import (CHUNK_COUNT, NameInfo, chunk_frames, go_frame,
                       is_ack, is_chunk, is_last_chunk, name_write_frame,
                       open_dump_req, open_write_frame, parse, pull_next_req,
                       read_name_req)
from .transport import Transport

_POLL_SLEEP = 0.002


class Session:
    """One serialized request/reply conversation with the device."""

    def __init__(self, transport: Transport, *,
                 name_timeout: float = 1.0, dump_timeout: float = 1.5,
                 ack_timeout: float = 1.0, name_retries: int = 3,
                 clock: Callable[[], float] = time.monotonic,
                 sleep: Callable[[float], None] = time.sleep):
        self.transport = transport
        self.name_timeout = name_timeout
        self.dump_timeout = dump_timeout
        self.ack_timeout = ack_timeout
        self.name_retries = name_retries
        self.clock = clock
        self.sleep = sleep
        self._lock = threading.Lock()
        self._seq = 0

    # ------------------------------------------------------------- public

    def read_name(self, slot: int) -> NameInfo:
        with self._lock:
            return self._transact_addressed(
                read_name_req(self._next_seq(), slot), slot)

    def read_blob(self, slot: int) -> bytes:
        with self._lock:
            return self._read_blob_locked(slot)

    def write_preset(self, slot: int, preset: Preset,
                     cancel: Optional[CancelToken] = None) -> NameInfo:
        """The gate-verified 7-frame write sequence, verbatim from
        docs/write-protocol.md. No checksum is computed at any point; none
        exists. The returned NameInfo is frame 7's read-back — comparison is
        the caller's job (the frame itself is protocol fidelity and always
        sent)."""
        frames = chunk_frames(preset.blob)   # validates blob size up front
        with self._lock:
            # 1. name read (fidelity to MCC; result discarded)
            self._transact_addressed(read_name_req(self._next_seq(), slot), slot)
            chunks_sent = 0
            stage = "name_write"
            try:
                # 2. name + meta — the device acks it with 0x18 (captured:
                #    every outbound 0x52 and 0x15 is acked, not just chunks)
                self._send(name_write_frame(self._next_seq(), slot,
                                            preset.name, preset.meta),
                           "name_write", slot, chunks_sent)
                self._require_ack("name_write", slot, chunks_sent)
                # 3. open blob write — acked with 0x18
                stage = "open"
                self._send(open_write_frame(self._next_seq(), slot),
                           "open", slot, chunks_sent)
                self._require_ack("open", slot, chunks_sent)
                # 4. go — acked with 0x18
                stage = "go"
                self._send(go_frame(), "go", slot, chunks_sent)
                self._require_ack("go", slot, chunks_sent)
                # 5-6. 146 chunks, each acked by the device with 0x18.
                # The three control-frame acks above are consumed before the
                # chunk loop, so ack N here really is the ack for chunk N —
                # flow control stays in lockstep with the device.
                stage = "chunk"
                for i, ch in enumerate(frames):
                    if cancel is not None and cancel.cancelled:
                        # Cancelling mid-write tears the slot; the fields say
                        # how far it got — recovery is "write again".
                        raise OperationCancelledError(i, len(frames))
                    self._send(ch, "chunk", slot, chunks_sent)
                    chunks_sent += 1
                    if not self._await_ack():
                        raise ChunkNotAckedError(slot, i)
            except (ChunkNotAckedError, OperationCancelledError,
                    WriteAbortedError):
                raise
            except (TransportError, ProtocolError) as e:
                raise WriteAbortedError(stage, slot, chunks_sent) from e
            # 7. name read back
            try:
                return self._transact_addressed(
                    read_name_req(self._next_seq(), slot), slot)
            except (DeviceTimeoutError, ReplyMismatchError, TransportError) as e:
                raise WriteAbortedError("final_read", slot, chunks_sent) from e

    def write_name(self, slot: int, name: str, meta: bytes) -> NameInfo:
        """Rename-in-place, exactly what MCC sends: the long 0x52 frame
        (acked by the device with 0x18, per the c4 capture) plus a refresh
        read. No blob traffic."""
        with self._lock:
            self.transport.send(name_write_frame(self._next_seq(), slot,
                                                 name, meta))
            if not self._await_ack():
                raise DeviceTimeoutError("name_write_ack", slot)
            return self._transact_addressed(
                read_name_req(self._next_seq(), slot), slot)

    def close(self) -> None:
        self.transport.close()

    # ------------------------------------------------------------ private

    def _next_seq(self) -> int:
        self._seq = self._seq % 127 + 1     # 1..127, never 0
        return self._seq

    def _send(self, message: bytes, stage: str, slot: int,
              chunks_sent: int) -> None:
        try:
            self.transport.send(message)
        except TransportError as e:
            raise WriteAbortedError(stage, slot, chunks_sent) from e

    def _require_ack(self, stage: str, slot: int, chunks_sent: int) -> None:
        """Await the device's 0x18 ack for a write control frame (long 0x52,
        short 0x52 open, go); absence aborts the write at that stage."""
        if not self._await_ack():
            raise WriteAbortedError(stage, slot, chunks_sent)

    def _drain(self) -> None:
        while self.transport.receive(0.0) is not None:
            pass

    def _next_frame(self, timeout: float) -> Optional[protocol.Frame]:
        """Next parseable MicroFreak frame within timeout, else None."""
        deadline = self.clock() + timeout
        while True:
            remaining = deadline - self.clock()
            if remaining <= 0:
                return None
            raw = self.transport.receive(remaining)
            if raw is None:
                self.sleep(_POLL_SLEEP)
                continue
            f = parse(raw)
            if f is not None:
                return f

    def _transact_addressed(self, request: bytes, slot: int) -> NameInfo:
        """Send an addressed request; accept only a long-0x52 reply whose
        embedded bank/pos names the requested slot — the reply-lag defense.

        A mismatched reply is discarded as stale and the request retried,
        name_retries times total; then ReplyMismatchError. Total silence
        raises DeviceTimeoutError(stage="name_read", slot=slot).

        Known limitation: when two consecutive reads target the SAME slot
        (write_preset frames 1 and 7; a rename's read-back), a lagged reply
        to the first is indistinguishable from the answer to the second —
        the embedded address matches either way. Blob-hash verification
        covers write(); for a rename this could in principle compare against
        a stale pre-rename name (whether a lagged hardware reply carries
        request-time or emission-time content is unobserved). No
        protocol-level fix exists: replies carry no request id.
        """
        self._drain()
        saw_reply = False
        replied_slot: Optional[int] = None
        for _attempt in range(self.name_retries):
            self.transport.send(request)
            deadline = self.clock() + self.name_timeout
            while True:
                remaining = deadline - self.clock()
                if remaining <= 0:
                    break                       # silent attempt; resend
                f = self._next_frame(remaining)
                if f is None:
                    break                       # silent attempt; resend
                if f.cmd != protocol.CMD_NAME or \
                        len(f.data) != protocol.NAME_PAYLOAD_LEN:
                    continue                    # unrelated traffic; ignore
                try:
                    info = protocol.decode_name_reply(f)
                except ProtocolError:
                    continue                    # malformed (e.g. out-of-range
                                                # embedded address); ignore
                saw_reply = True
                if info.slot == slot:
                    return info
                replied_slot = info.slot        # stale (lagged) reply
                break                           # discard and resend at once
        if saw_reply:
            raise ReplyMismatchError(slot, replied_slot, self.name_retries)
        raise DeviceTimeoutError("name_read", slot)

    def _read_blob_locked(self, slot: int) -> bytes:
        """Open a dump, then pull chunks in strict lockstep — one outstanding
        request — until the device says 'last'."""
        self._drain()
        self.transport.send(open_dump_req(self._next_seq(), slot))
        chunks: List[protocol.Frame] = []
        while True:
            # collect anything already delivered before pulling again; the
            # len bound keeps a runaway device (streaming 0x16 forever with
            # no terminator) from spinning this loop unboundedly
            got_last = False
            while len(chunks) < CHUNK_COUNT:
                raw = self.transport.receive(0.0)
                if raw is None:
                    break
                f = parse(raw)
                if f is None or not is_chunk(f):
                    continue
                chunks.append(f)
                if is_last_chunk(f):
                    got_last = True
                    break
            if got_last:
                return protocol.assemble_blob(chunks)
            if len(chunks) >= CHUNK_COUNT:
                raise ProtocolError(
                    f"slot {slot}: {len(chunks)} chunks without a 0x17 terminator")
            self.transport.send(pull_next_req(self._next_seq()))
            f = self._await_chunk(slot)
            chunks.append(f)
            if is_last_chunk(f):
                return protocol.assemble_blob(chunks)

    def _await_chunk(self, slot: int) -> protocol.Frame:
        deadline = self.clock() + self.dump_timeout
        while True:
            remaining = deadline - self.clock()
            if remaining <= 0:
                raise DeviceTimeoutError("dump", slot)
            f = self._next_frame(remaining)
            if f is None:
                raise DeviceTimeoutError("dump", slot)
            if is_chunk(f):
                return f
            # anything else during a dump is stale; discard

    def _await_ack(self) -> bool:
        deadline = self.clock() + self.ack_timeout
        while True:
            remaining = deadline - self.clock()
            if remaining <= 0:
                return False
            f = self._next_frame(remaining)
            if f is None:
                return False
            if is_ack(f):
                return True
            # anything else during a write is stale; discard
